ulid.go

```diff
@@ -454,3 +461,40 @@ func (id *ULID) Scan(src interface{}) error {
 func (id ULID) Value() (driver.Value, error) {
 	return id.MarshalBinary()
 }
+
+// Monotonic returns an entropy source that is guaranteed to return
+// strictly increasing entropy bytes for the same ULID timestamp.
+// Not safe for concurrent use.
+func Monotonic(entropy io.Reader) io.Reader {
+	return &monotonic{Reader: entropy}
+}
+
+type monotonic struct {
+	io.Reader
+	ms      uint64
+	entropy []byte
+}
+
+func (m *monotonic) MonotonicRead(ms uint64, entropy []byte) (err error) {
+	if len(m.entropy) != 0 && m.ms == ms {
+		err = m.increment()
+		copy(entropy, m.entropy)
+	} else if _, err = io.ReadFull(m, entropy); err == nil {
+		m.ms = ms
+		m.entropy = append(m.entropy[:0], entropy...)
+	}
+	return err
+}
+
+func (m *monotonic) increment() error {
+	for i := len(m.entropy) - 1; i >= 0; i-- {
+		switch m.entropy[i] {
+		case 255:
+			m.entropy[i] = 0
+		default:
+			m.entropy[i]++
+			return nil
+		}
```

Comment:
Cute. Did you take this from somewhere?

Reply:
From here: https://github.com/oklog/ulid/pull/15/files#diff-ccc49aab1ea1bfc1c623feb8fb450309R39